### PR TITLE
feat: Improve code detection for UI5 classes (JavaScript)

### PR DESCRIPTION
**Pull request by matz3**
This change improves the detection of JavaScript code related to UI5 classes written in the scanned project by converting all UI5 classes to ES classes.

Previously, the linter only converted some of the UI5 classes which left some code uncovered as the linter did not recognize the proper types.

This is also a preparation to enable further checks on sub-classes of UI5 framework base classes such as controls and elements.

### CODE REVIEW

File: test/fixtures/transpiler/amd/Factory_BasicModule.js [hunk 1]
```diff
@@ -2,6 +2,9 @@ sap.ui.define(["sap/ui/core/mvc/Controller", "sap/ui/core/UIComponent", "sap/ui/
 	function (Controller, UIComponent, History, Button) {
 	"use strict";
 
+	/**
+	 * @class JSDoc comments should be removed, as it might have negative impact on the type detection
+	 */
```
Comment (d3xter666) [inline test/fixtures/transpiler/amd/Factory_BasicModule.js:7]: If TS parses and extracts comments the way JSDco does, then this comment would be totally ignored. JSDoc ignores comments above return statements

File: src/linter/ui5Types/amdTranspiler/insertNodesInParent.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * Inserts a node after the specified node in the parent block
+ */
+export default function insertNodesInParent(
+	parentNode: ts.SourceFile | ts.Block, insertionsMap: Map<ts.Statement, ts.Statement[]>, nodeFactory: ts.NodeFactory
+): ts.SourceFile | ts.Block | undefined {
+	const insertOperations: InsertOperation[] = [];
+	for (const [referenceNode, nodesToBeInserted] of insertionsMap) {
+		const index = parentNode.statements.indexOf(referenceNode);
+		if (index != -1) {
```
Comment (d3xter666) [inline src/linter/ui5Types/amdTranspiler/insertNodesInParent.ts:17]:
```suggestion
		if (index !== -1) {
```

**Review by d3xter666 — COMMENTED** [review]
LGTM

File: test/fixtures/transpiler/amd/Factory_BasicModule.js (hunk 1 above)
Comment (matz3) [reply test/fixtures/transpiler/amd/Factory_BasicModule.js:7]: That's true. Right now we ignore JSDoc comments everywhere, but in this case it wouldn't matter.

File: src/linter/ui5Types/amdTranspiler/insertNodesInParent.ts (hunk 2 above)
Comment (matz3) [reply src/linter/ui5Types/amdTranspiler/insertNodesInParent.ts]: Done

**Review by d3xter666 — DISMISSED** [review]

**Review by maxreichmann — DISMISSED** [review]

File: src/linter/ui5Types/amdTranspiler/tsTransformer.ts [hunk 3]
```diff
@@ -112,6 +144,55 @@ function transform(
...
+					) {
+						variableStatement = node.parent.parent.parent;
+						className = node.parent.name.getText();
+					}
+
+					// For now, only rewrite extend calls in expressions and variable statements
+					if (variableStatement || ts.isExpressionStatement(node.parent)) {
+						const classDeclaration = rewriteExtendCall(nodeFactory, node, undefined, className);
+						if (variableStatement) {
+							// We can't replace the variable declaration with the class declaration (not valid),
```
Comment (RandomByte) [inline src/linter/ui5Types/amdTranspiler/tsTransformer.ts:167]: Should be moved into the next if, right? I.e. it doesn't apply to the else case

**Review by RandomByte — COMMENTED** [review]
Looks good to me!

File: src/linter/ui5Types/amdTranspiler/tsTransformer.ts (hunk 3 above)
Comment (matz3) [reply src/linter/ui5Types/amdTranspiler/tsTransformer.ts]: True, thanks. I've update the comments.

File: src/linter/ui5Types/amdTranspiler/tsTransformer.ts [hunk 4]
```diff
@@ -57,6 +72,23 @@ function transform(
 		replacements.push({original: node, substitute});
 	}
 
+	function insertNodeAfter(referenceNode: ts.Statement, nodeToBeInserted: ts.Statement) {
+		if (!ts.isBlock(referenceNode.parent)) {
+			return;
```
[the remark this answers is not in the document]
Comment (matz3) [reply src/linter/ui5Types/amdTranspiler/tsTransformer.ts]: Done. I extended the logic to support BlockLike, which should cover all possible cases where statements can be inserted.

File: src/linter/ui5Types/amdTranspiler/tsTransformer.ts [hunk 5]
```diff
@@ -112,6 +144,55 @@ function transform(
...
+							if (variableStatement.declarationList.declarations.length > 1) {
+								// The variable statement contains more than just our class variable,
+								// so we just remove the single declaration within it
+								pruneNode(node.parent);
+
+								// And insert the class declaration after the variable statement
+								insertNodeAfter(variableStatement, classDeclaration);
+
+								// Also: Move comments from the variable declaration to the class declaration
+								moveCommentsToNode(node.parent, classDeclaration, sourceFile);
```
[the remark this answers is not in the document]
Comment (matz3) [reply src/linter/ui5Types/amdTranspiler/tsTransformer.ts:185]: Yes, there are some tricky cases. As we don't need to comments as of today, let's keep it as it is and improve this later.

**Review by RandomByte — APPROVED** [review]

**Pull request merged**